### PR TITLE
OpenAPI built-in compiler extension implementation

### PR DESCRIPTION
**Pull request by ayeshLK**
## Purpose
As an effort to support REST API development we are implementing an introspection endpoint for any ballerina HTTP service and that particular introspection endpoint will be serving a OpenAPI Doc for that particular service.

## Implementation
We have implement this feature via in-built compiler plugin.
- Identify HTTP service declaration.
- Check whether `@openapi:ServiceInfo` is present with valid `contract` file, if so pack that contract file into ballerina executable jar.
- If `@openapi:ServiceInfo` is not present, then generate a OpenAPI doc for the service declaration and pack it into ballerina executable jar.

When packing the OpenAPI doc as a resource to ballerina executable jar we would use following structure. (Please refer [this](https://github.com/ballerina-platform/ballerina-standard-library/issues/1776#issuecomment-901665069))
```
├── target/bin
│   ├-- 1projectapi.jar
│   │   └── resources
│   │       └── ballerina
│   │           └── http
│   │               └── ry328y8239yr28d8dheuhnxh.json
```
Generated OpenAPI doc will be named in following format.
```
<hashCode-for-ServiceDeclarationSymbol>.json
```

## Limitations
Currently this will only work for ballerina projects.

Related to [#1616](https://github.com/ballerina-platform/ballerina-standard-library/issues/1616)
Related to [#1776](https://github.com/ballerina-platform/ballerina-standard-library/issues/1776)

### CODE REVIEW

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Writes a file with content to specified {@code filePath}.
+     *
+     * @param filePath  valid file path to write the content
+     * @param content  content of the file
+     * @param append  flag to notify whether to append content or not
+     * @throws IOException when a file operation fails
+     */
+    public static void writeFile(Path filePath, String content, boolean append) throws IOException {
+        try (FileWriter writer = new FileWriter(filePath.toString(), StandardCharsets.UTF_8, append)) {
```
Comment (chamil321) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java:51]: Do we require this `append` argument. by default it's false for the `FileWriter` method. And I dont see any usage of true being pass for `append` argument.

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Copy content of a file/directory into another location.
+     *
+     * @param inputStream  stream from which the data is read
+     * @param outStream  stream to which the data is written
+     * @throws IOException if there is any error while reading from a file or writing to a file
+     */
+    public static <T extends InputStream, E extends OutputStream> void copyContent(T inputStream, E outStream)
+            throws IOException {
```
Comment (chamil321) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java:83]: Where do we catch this Exception?

Comment (ayeshLK) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java:83]: This exception is caught in the code block which uses this function.

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (lnash94) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/utils/CodegenUtils.java:2]: Shall we change the license header to 2021?

File: openapi-extension-tests/src/test/resources/ballerina_sources/sample_4/service_openapi.json [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  },
+  "components" : { }
+}
```
Comment (lnash94) [inline openapi-extension-tests/src/test/resources/ballerina_sources/sample_4/service_openapi.json:30]: Shall we add the new line there, and other required places :) ?

File: openapi-extension-tests/src/test/java/io/ballerina/openapi/extension/OpenApiExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        PackageCompilation compilation = currentPackage.getCompilation();
+        testGeneratedResources("sample_2", 2);
+    }
+
+    @Test
+    public void testDocGenerationForBallerinaProjectWithMultipleModules() throws IOException {
+        deleteTarget("sample_3");
+        Package currentPackage = loadPackage("sample_3");
+        PackageCompilation compilation = currentPackage.getCompilation();
+//        testGeneratedResources("sample_3", 3);
```
Comment (shafreenAnfar) [inline openapi-extension-tests/src/test/java/io/ballerina/openapi/extension/OpenApiExtensionTest.java:67]: Any reason for this ?

Comment (ayeshLK) [reply openapi-extension-tests/src/test/java/io/ballerina/openapi/extension/OpenApiExtensionTest.java:67]: Currently for two service declarations in different ballerina source files will get same hash-code, hence this test case is not passing. Will un-comment this once lang fix is merged.

**Review by chamil321 — APPROVED** [review]

File: openapi-extension-tests/src/test/java/io/ballerina/openapi/extension/OpenApiExtensionTest.java (hunk 5 above)
Comment (ayeshLK) [reply openapi-extension-tests/src/test/java/io/ballerina/openapi/extension/OpenApiExtensionTest.java]: Will update the test-cases after RC2 pack is released.

File: openapi-extension/src/main/java/io/ballerina/openapi/extension/OpenApiDocPackagingTask.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        Path targetBinPath = executableJarAbsPath.getParent();
+        if (null != targetBinPath && Files.exists(targetBinPath)) {
+            String executableJarFileName = executableJarAbsPath.toFile().getName();
+            try {
+                this.packagingService.updateExecutableJar(targetBinPath, executableJarFileName);
+            } catch (IOException e) {
+                ERR.println("error [open-api extension]: " + e.getLocalizedMessage());
+            }
+
+            // clean up created intermediate resources if current project is a single-ballerina-file  project
```
Comment (ayeshLK) [inline openapi-extension/src/main/java/io/ballerina/openapi/extension/OpenApiDocPackagingTask.java:63]:
```suggestion
            // clean up created intermediate resources if current project is a single-ballerina-file project
```

**Pull request merged**